README.md
[![unit-tests-linux](https://github.com/drgfreeman/dynamo-pandas/actions/workflows/linux-checks.yml/badge.svg)](https://github.com/DrGFreeman/dynamo-pandas/actions/workflows/linux-checks.yml)
[![unit-tests-windows](https://github.com/drgfreeman/dynamo-pandas/actions/workflows/windows-checks.yml/badge.svg)](https://github.com/DrGFreeman/dynamo-pandas/actions/workflows/windows-checks.yml)
[![Documentation Status](https://readthedocs.org/projects/dynamo-pandas/badge/?version=latest)](https://dynamo-pandas.readthedocs.io/en/latest/?badge=latest)

# dynamo-pandas
Make working with pandas data and AWS DynamoDB easy.

## Motivation
This package aims a making the transfer of data between pandas dataframes and DynamoDB as simple as possible. To meet this goal, the package offers two key features:
1. Automatic conversion of pandas data types to DynamoDB supported data types.
1. A simple, high level interface to *put* data from a dataframe into a DynamoDB table and *get* all or selected items from a table into a dataframe.


## Documentation

The project's documentation is available at https://dynamo-pandas.readthedocs.io/.


## Requirements
* `python>=3.7`
* `pandas>=1`
* `boto3`

## Installation

```
python -m pip install dynamo-pandas
```

This will install the package and its dependencies except for `boto3` which is not installed by default to avoid unnecessary installation when building Lambda layers.

To include `boto3` as part of the installation, add the `boto3` "extra" this way:

```
python -m pip install dynamo-pandas[boto3]
```

## Example Usage

Consider the pandas DataFrame below.


```python
>>> print(players_df)

      player_id           last_play       play_time  rating  bonus_points
0    player_one 2021-01-18 22:47:23 2 days 17:41:55     4.3             3
1    player_two 2021-01-19 19:07:54 0 days 22:07:34     3.8             1
2  player_three 2021-01-21 10:22:43 1 days 14:01:19     2.5             4
3   player_four 2021-01-22 13:51:12 0 days 03:45:49     4.8          <NA>
```

The columns of the dataframe use different data types, some of which are not natively supported by DynamoDB, like numpy.datetime64, timedelta64 and pandas' nullable integers.


```python
>>> players_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4 entries, 0 to 3
Data columns (total 5 columns):
    #   Column        Non-Null Count  Dtype          
---  ------        --------------  -----          
    0   player_id     4 non-null      object         
    1   last_play     4 non-null      datetime64[ns] 
    2   play_time     4 non-null      timedelta64[ns]
    3   rating        4 non-null      float64        
    4   bonus_points  3 non-null      Int8           
dtypes: Int8(1), datetime64[ns](1), float64(1), object(1), timedelta64[ns](1)
memory usage: 264.0+ bytes
```

Storing the rows of this dataframe to DynamoDB requires multiple data type conversions.

```python
>>> from dynamo_pandas import put_df, get_df, keys
```

The `put_df` function adds or updates the rows of a dataframe into the specified table, taking care of the required type conversions (the table must be already created and the primary key column(s) be present in the dataframe).

```python
>>> put_df(players_df, table="players")
```

The `get_df` function retrieves the items matching the speficied key(s) from the table into a dataframe.


```python
>>> df = get_df(table="players", keys=[{"player_id": "player_three"}, {"player_id": "player_one"}])
>>> print(df)

   bonus_points     player_id            last_play  rating        play_time
0             4  player_three  2021-01-21 10:22:43     2.5  1 days 14:01:19
1             3    player_one  2021-01-18 22:47:23     4.3  2 days 17:41:55
```

In the case where only a partition key is used, the `keys` function simplifies the generation of the keys list.


```python
>>> df = get_df(table="players", keys=keys(player_id=["player_two", "player_four"]))
>>> print(df)

   bonus_points    player_id            last_play  rating        play_time
0           1.0   player_two  2021-01-19 19:07:54     3.8  0 days 22:07:34
1           NaN  player_four  2021-01-22 13:51:12     4.8  0 days 03:45:49
```

The data types returned by the `get_df` function are basic types and no automatic type conversion is attempted.


```python
>>> df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2 entries, 0 to 1
Data columns (total 5 columns):
    #   Column        Non-Null Count  Dtype  
   ---  ------        --------------  -----  
    0   bonus_points  1 non-null      float64
    1   player_id     2 non-null      object 
    2   last_play     2 non-null      object 
    3   rating        2 non-null      float64
    4   play_time     2 non-null      object 
dtypes: float64(2), object(3)
memory usage: 208.0+ bytes
```

The `dtype` parameter of the `get_df` function allows specifying the desired data types.

```python
>>> df = get_df(
...     table="players",
...     keys=keys(player_id=["player_two", "player_four"]),
...     dtype={
...         "bonus_points": "Int8",
...         "last_play": "datetime64[ns, UTC]",
...         # "play_time": "timedelta64[ns]"  # See note below.
...     }
... )
```

**Note**: Due to a known bug in pandas, timedelta strings cannot currently be converted back to Timedelta type via this parameter (ref. https://github.com/pandas-dev/pandas/issues/38509). Use the pandas.to_timedelta function instead:


```python
>>> df.play_time = pd.to_timedelta(df.play_time)
>>> df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2 entries, 0 to 1
Data columns (total 5 columns):
    #   Column        Non-Null Count  Dtype              
---  ------        --------------  -----              
    0   bonus_points  1 non-null      Int8               
    1   player_id     2 non-null      object             
    2   last_play     2 non-null      datetime64[ns, UTC]
    3   rating        2 non-null      float64            
    4   play_time     2 non-null      timedelta64[ns]    
dtypes: Int8(1), datetime64[ns, UTC](1), float64(1), object(1), timedelta64[ns](1)
memory usage: 196.0+ bytes
```

Omitting the `keys` parameter performs a scan of the table and returns all the items.


```python
>>> df = get_df(table="players")
>>> print(df)

       bonus_points     player_id            last_play  rating        play_time
    0           4.0  player_three  2021-01-21 10:22:43     2.5  1 days 14:01:19
    1           NaN   player_four  2021-01-22 13:51:12     4.8  0 days 03:45:49
    2           3.0    player_one  2021-01-18 22:47:23     4.3  2 days 17:41:55
    3           1.0    player_two  2021-01-19 19:07:54     3.8  0 days 22:07:34
```

## License

Released under the terms of the [MIT License](LICENSE).

CHANGELOG.md
## Unreleased

* ...

## Version 1.3.0

### New Features

* Allow passing boto3 resource or client parameters to functions (issue #52, PR #57).

### Fixes

* Fix indentation in docs code examples (issues #50, #55, PRs #53, #56).

### Other changes

* Add testing under Python 3.10 (issue #58, PR #59).

## Version 1.2.1

### Fixes

* Fix typo in `transactions.put_items` unprocessed items key (issue #42, PR #45).
* Fix handling of unprocessed items in `transactions.put_items` (issue #44, PR #45).
* Fix handling of unprocessed keys in `transactions.get_items` (issue #46, PR #45).

### Other changes

* Add unit testing of unprocessed items/keys in `transactions.put_items` and `transactions.get_items` (issue #43, PR #45).
* Add missing lines to coverage report in tox config (PR #45).

## Version 1.2.0

### New Features

* Add the `attributes` parameter to the `get_df`, `transactions.get_item`, `transactions.get_items` and `transactions.get_all_items` functions. (issue #39, PR #40)


## Version 1.1.1

### Fixes

* Fix version number which was not updated in version 1.1.0.


## Version 1.1.0

### Modified Features

* Set `boto3` as an optional requirement to avoid unnecessary installation in lambda layers (issue #35, PR #36).

tests/conftest.py
import base64
import os

import boto3
from moto import mock_dynamodb
import pytest
from test_data import large_table_items
from test_data import test_df

from dynamo_pandas.transactions import put_item


@pytest.fixture()
def aws_credentials():
    """Dummy AWS credentials to use with the moto mocks."""
    os.environ["AWS_ACCESS_KEY_ID"] = "testing"
    os.environ["AWS_SECRET_ACCESS_KEY"] = "testing"
    os.environ["AWS_SECURITY_TOKEN"] = "testing"
    os.environ["AWS_SESSION_TOKEN"] = "testing"
    os.environ["AWS_DEFAULT_REGION"] = "us-east-1"


@pytest.fixture()
def ddb_client(aws_credentials):
    """Fixture to mock the dynamodb client using moto."""
    with mock_dynamodb():
        yield boto3.client("dynamodb")


@pytest.fixture()
def empty_table(ddb_client):
    """Fixture generating an empty table and yielding the name of the table. The table
    primary key is named 'id' and is of numerical type."""
    table_name = "test-empty-table"
    response = ddb_client.create_table(
        AttributeDefinitions=[dict(AttributeName="id", AttributeType="N")],
        TableName=table_name,
        KeySchema=[dict(AttributeName="id", KeyType="HASH")],
        BillingMode="PROVISIONED",
        ProvisionedThroughput=dict(ReadCapacityUnits=5, WriteCapacityUnits=5),
    )
    assert response["ResponseMetadata"]["HTTPStatusCode"] == 200
    yield table_name


@pytest.fixture()
def test_df_table(ddb_client):
    """Fixture generating a table with the items from test_df and yielding the name of
    the table. The table primary key is named 'id' and is of numerical type."""
    table_name = "test-df-table"
    response = ddb_client.create_table(
        AttributeDefinitions=[dict(AttributeName="id", AttributeType="N")],
        TableName=table_name,
        KeySchema=[dict(AttributeName="id", KeyType="HASH")],
        BillingMode="PROVISIONED",
        ProvisionedThroughput=dict(ReadCapacityUnits=5, WriteCapacityUnits=5),
    )
    assert response["ResponseMetadata"]["HTTPStatusCode"] == 200

    for item in test_df.to_dict("records"):
        response = put_item(item=item, table=table_name, return_response=True)
        assert response["ResponseMetadata"]["HTTPStatusCode"] == 200

    yield table_name


@pytest.fixture()
def large_table(ddb_client):
    """Fixture providing a large table (250 items) to test get operations with more
    than 100 items. The table primary key is named 'id' and is of numerical type."""
    table_name = "test-large-table"
    response = ddb_client.create_table(
        AttributeDefinitions=[dict(AttributeName="id", AttributeType="N")],
        TableName=table_name,
        KeySchema=[dict(AttributeName="id", KeyType="HASH")],
        BillingMode="PROVISIONED",
        ProvisionedThroughput=dict(ReadCapacityUnits=5, WriteCapacityUnits=5),
    )
    assert response["ResponseMetadata"]["HTTPStatusCode"] == 200

    for item in large_table_items:
        response = put_item(item=item, table=table_name, return_response=True)
        assert response["ResponseMetadata"]["HTTPStatusCode"] == 200

    yield table_name


@pytest.fixture()
def large_objects_table(ddb_client):
    """Fixture providing a table with large objects (~395 KB each) to test get
    operations with more generating large data sizes. The table primary key is named
    'id' and is of numerical type."""
    table_name = "test-large-objects-table"
    response = ddb_client.create_table(
        AttributeDefinitions=[dict(AttributeName="id", AttributeType="N")],
        TableName=table_name,
        KeySchema=[dict(AttributeName="id", KeyType="HASH")],
        BillingMode="PAY_PER_REQUEST",
    )
    assert response["ResponseMetadata"]["HTTPStatusCode"] == 200

    for i in range(100):
        item = dict(id=i, A=base64.b64encode(os.urandom(296 * 1024)).decode())
        response = put_item(item=item, table=table_name, return_response=True)
        assert response["ResponseMetadata"]["HTTPStatusCode"] == 200

    yield table_name


tests/test_data.py
import math
import random
from string import ascii_letters

import pandas as pd

"""Define a test dataframe with mixed types and missing values:
>>> print(test_df)
      A  B               C          D                         E     F  \
0   abc  2 0 days 19:32:01 2000-01-01 2000-01-01 00:00:00+00:00   128   
1  None  3 1 days 01:33:20 2000-12-31 2000-12-31 23:59:59+00:00  <NA>   
2   NaN  4 2 days 23:06:40        NaT                       NaT  <NA>   

          G  id  
0  3.141593   0  
1       NaN   1  
2       NaN   2  
>>> test_df.info()
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3 entries, 0 to 2
Data columns (total 8 columns):
 #   Column  Non-Null Count  Dtype              
---  ------  --------------  -----              
 0   A       1 non-null      object             
 1   B       3 non-null      int64              
 2   C       3 non-null      timedelta64[ns]    
 3   D       2 non-null      datetime64[ns]     
 4   E       2 non-null      datetime64[ns, UTC]
 5   F       1 non-null      Int32              
 6   G       1 non-null      float64            
 7   id      3 non-null      int64              
dtypes: Int32(1), datetime64[ns, UTC](1), datetime64[ns](1), float64(1), int64(2), object(1), timedelta64[ns](1)
memory usage: 311.0+ bytes"""  # noqa: E501, W291
test_df = pd.DataFrame(
    [
        {
            "A": "abc",
            "B": 2,
            "C": 70321.4,
            "D": "2000-01-01",
            "E": "2000-01-01",
            "F": 128,
            "G": math.pi,
        },
        {
            "A": None,
            "B": 3,
            "C": 92000,
            "D": "2000-12-31",
            "E": "2000-12-31 23:59:59",
            "F": None,
            "G": None,
        },
        {"B": 4, "C": 256000},
    ]
).astype(
    {
        "C": "timedelta64[s]",
        "D": "datetime64[ns]",
        "E": "datetime64[ns, UTC]",
        "F": "Int32",
    }
)
test_df["id"] = range(len(test_df))

large_table_items = [
    dict(id=i, letter=random.choice(ascii_letters), number=random.randint(0, 1000))
    for i in range(250)
]


tests/test_main.py
import re
from unittest import mock

import pandas as pd
import pytest
from test_data import test_df

from dynamo_pandas import get_df
from dynamo_pandas import keys
from dynamo_pandas import put_df
from dynamo_pandas.dynamo_pandas import _to_df
from dynamo_pandas.dynamo_pandas import _to_items

# List of item dictionaries with pandas dtypes
test_items_pd = test_df.to_dict("records")

# List of item dictionaries with types as returned from DyanamoDB when using the
# functions in the transactions module.
test_items = test_df.astype(dict(C="str", D="str", E="str", F="float")).to_dict(
    "records"
)


class Test_keys:
    """Test the keys function."""

    def test_partition_key_only(self):
        """Test that the keys function works with only a partition key."""
        assert keys(id=range(3)) == [{"id": 0}, {"id": 1}, {"id": 2}]

    def test_two_or_more_kwargs_raises(self):
        """Test that two or more keyword arguments raises a ValueError."""
        with pytest.raises(
            ValueError,
            match=re.escape("Only one key attribute (partition key) is supported."),
        ):
            keys(id=[1, 2, 3], di=[3, 2, 1])


class Test_get_df:
    """Test the get_df function."""

    def test_no_keys(self, test_df_table):
        """Test that not specifying keys returns all items."""
        df = get_df(table=test_df_table)

        assert {c: t.name for c, t in zip(df.columns, df.dtypes)} == {
            "A": "object",
            "B": "int64",
            "C": "object",
            "D": "object",
            "E": "object",
            "F": "float64",
            "G": "float64",
            "id": "int64",
        }
        assert df.equals(
            pd.DataFrame(
                [
                    {
                        "A": "abc",
                        "B": 2,
                        "C": "0 days 19:32:01",
                        "D": "2000-01-01 00:00:00",
                        "E": "2000-01-01 00:00:00+00:00",
                        "F": 128.0,
                        "G": 3.141592653589793,
                        "id": 0,
                    },
                    {
                        "A": None,
                        "B": 3,
                        "C": "1 days 01:33:20",
                        "D": "2000-12-31 00:00:00",
                        "E": "2000-12-31 23:59:59+00:00",
                        "id": 1,
                    },
                    {
                        "A": None,
                        "B": 4,
                        "C": "2 days 23:06:40",
                        "D": None,
                        "E": None,
                        "id": 2,
                    },
                ]
            )
        )

    def test_keys(self, test_df_table):
        """Test with keys specified. Also test that keys not in the table are
        ignored."""
        df = get_df(table=test_df_table, keys=[{"id": 0}, {"id": 2}, {"id": 3}])

        assert {c: t.name for c, t in zip(df.columns, df.dtypes)} == {
            "A": "object",
            "B": "int64",
            "C": "object",
            "D": "object",
            "E": "object",
            "F": "float64",
            "G": "float64",
            "id": "int64",
        }
        assert df.equals(
            pd.DataFrame(
                [
                    {
                        "A": "abc",
                        "B": 2,
                        "C": "0 days 19:32:01",
                        "D": "2000-01-01 00:00:00",
                        "E": "2000-01-01 00:00:00+00:00",
                        "F": 128.0,
                        "G": 3.141592653589793,
                        "id": 0,
                    },
                    {
                        "A": None,
                        "B": 4,
                        "C": "2 days 23:06:40",
                        "D": None,
                        "E": None,
                        "id": 2,
                    },
                ]
            )
        )

    def test_single_key_missing(self, test_df_table):
        """Test that a single key not in the table returns an empty dataframe."""
        df = get_df(table=test_df_table, keys=[dict(id=3)])

        assert df.empty

    def test_attributes_returns_specified_columns(self, test_df_table):
        """Test that only columns corresponding to the specified attributes are
        returned."""
        df = get_df(
            table=test_df_table,
            attributes=["id", "A", "E"],
            keys=[{"id": 0}, {"id": 1}],
        )

        assert df.equals(
            pd.DataFrame(
                [
                    {"A": "abc", "E": "2000-01-01 00:00:00+00:00", "id": 0},
                    {"A": None, "E": "2000-12-31 23:59:59+00:00", "id": 1},
                ]
            )
        )

    def test_attributes_and_keys(self, test_df_table):
        """Test that only columns corresponding to the specified attributes are
        returned along with keys."""
        df = get_df(table=test_df_table, attributes=["id", "A", "E"])

        assert df.equals(
            pd.DataFrame(
                [
                    {"A": "abc", "E": "2000-01-01 00:00:00+00:00", "id": 0},
                    {"A": None, "E": "2000-12-31 23:59:59+00:00", "id": 1},
                    {"A": None, "E": None, "id": 2},
                ]
            )
        )

    def test_dtype(self, test_df_table):
        """Test that the dtype parameter controls the returned data types."""
        df = get_df(
            table=test_df_table,
            dtype={
                # "C": "timedelta64[ns]",  # Ref. #24
                "D": "datetime64[ns]",
                "E": "datetime64[ns, UTC]",
                "F": "Int32",
            },
        )
        assert {c: t.name for c, t in zip(df.columns, df.dtypes)} == {
            "A": "object",
            "B": "int64",
            "C": "object",
            "D": "datetime64[ns]",
            "E": "datetime64[ns, UTC]",
            "F": "Int32",
            "G": "float64",
            "id": "int64",
        }

    @pytest.mark.parametrize("keys", (None, [{"id": 0}]))
    def test_boto3_kwargs_are_passed(self, ddb_client, test_df_table, keys):
        """Test that the boto3_kwargs are passed to the boto3.resource() function
        call."""
        # test_df_table is defined in us-east-1. By setting the region_name to
        # ca-central-1, we expect a ResourceNotFoundException.
        # Moto raises a ValueError instead so expect it as well
        # (see https://github.com/spulec/moto/issues/4344)
        with pytest.raises(
            (ddb_client.exceptions.ResourceNotFoundException, ValueError)
        ):
            get_df(
                keys=keys,
                table=test_df_table,
                boto3_kwargs=dict(region_name="ca-central-1"),
            )

        # With the correct region we expect to get the items.
        df = get_df(
            keys=keys,
            table=test_df_table,
            boto3_kwargs=dict(region_name="us-east-1"),
        )
        assert not df.empty


class Test_put_df:
    """Test the put_df function."""

    def test_with_pd_types(self, ddb_client, empty_table):
        """Test putting a dataframe using pandas specific data types."""
        put_df(df=test_df, table=empty_table)

        assert get_df(table=empty_table).equals(
            pd.DataFrame(
                [
                    {
                        "A": "abc",
                        "B": 2,
                        "C": "0 days 19:32:01",
                        "D": "2000-01-01 00:00:00",
                        "E": "2000-01-01 00:00:00+00:00",
                        "F": 128.0,
                        "G": 3.141592653589793,
                        "id": 0,
                    },
                    {
                        "A": None,
                        "B": 3,
                        "C": "1 days 01:33:20",
                        "D": "2000-12-31 00:00:00",
                        "E": "2000-12-31 23:59:59+00:00",
                        "id": 1,
                    },
                    {
                        "A": None,
                        "B": 4,
                        "C": "2 days 23:06:40",
                        "D": None,
                        "E": None,
                        "id": 2,
                    },
                ]
            )
        )

    def test_boto3_kwargs_are_passed(self, ddb_client, empty_table):
        """Test that the boto3_kwargs are passed to the boto3.client() function call."""
        # Moto does not raise the expected ResourceNotFoundError (see
        # https://github.com/spulec/moto/issues/4347)
        # We mock the boto3.client call instead and verify the boto3_kwargs are passed
        with mock.patch(
            "dynamo_pandas.transactions.transactions.boto3.client"
        ) as client:
            put_df(
                test_df,
                table=empty_table,
                boto3_kwargs=dict(region_name="ca-central-1"),
            )

            assert client.call_args[1] == dict(region_name="ca-central-1")


class Test__to_df:
    """Test the _to_df function."""

    def test_single_item_dict(self):
        """Test conversion of a single item dictionary."""
        df = _to_df(test_items[0])

        assert df.equals(pd.DataFrame([test_items[0]]))
        assert [t.name for t in df.dtypes] == [
            "object",
            "int64",
            "object",
            "object",
            "object",
            "float64",
            "float64",
            "int64",
        ]

    def test_multiple_item_dicts(self):
        """Test conversion of a list of item dictionaries."""
        df = _to_df(test_items)

        assert df.equals(pd.DataFrame(test_items))
        assert [t.name for t in df.dtypes] == [
            "object",
            "int64",
            "object",
            "object",
            "object",
            "float64",
            "float64",
            "int64",
        ]

    def test_with_dtype(self):
        """Test with dtype parameter specified."""
        df = _to_df(
            test_items,
            dtype=dict(
                # C="timedelta64[ns]",  # Ref. #24
                D="datetime64",
                E="datetime64[ns, UTC]",
                F="Int32",
            ),
        )

        assert [t.name for t in df.dtypes] == [
            "object",
            "int64",
            "object",
            "datetime64[ns]",
            "datetime64[ns, UTC]",
            "Int32",
            "float64",
            "int64",
        ]


class Test__to_items:
    """Test the _to_items function."""

    def test_df(self):
        """Test that a dataframe converts to a list of item dictionaries."""
        items = _to_items(test_df)

        # Compare dictionary string values as pd.NA and np.nan fail comparison
        assert str(items) == str(test_items_pd)

    def test_invalid_type_raises(self):
        """Test that a type different than a DataFrame raises a TypeError."""
        with pytest.raises(TypeError, match="df must be a pandas DataFrame"):
            _to_items(test_items_pd)


dynamo_pandas/dynamo_pandas.py
import pandas as pd

from .transactions import get_all_items
from .transactions import get_items
from .transactions import put_items


def get_df(*, table, keys=None, attributes=None, dtype=None, boto3_kwargs={}):
    """Get items from a table into a dataframe.

    Parameters
    ----------
    table : str
        Name of the DynamoDB table.

    keys : list[dict]
        List of keys to get where each key is represented by a dictionary.

    attributes : list[str]
        Names of the item attributes to return as dataframe columns. If None (default),
        all attributes are returned.

    dtype : data type or dict of column names -> data type
        Use a numpy.dtype or Python type to cast entire pandas object to the same type.
        Alternatively, use {col: dtype, …}, where col is a column label and dtype is a
        numpy.dtype or Python type to cast one or more of the DataFrame’s columns to
        column-specific types.

    boto3_kwargs : dict
        Keyword arguments to pass to the underlying ``boto3.resource('dynamodb')``
        function call (see `boto3 docs <https://boto3.amazonaws.com/v1/documentation/api/latest/reference/core/session.html#boto3.session.Session.resource>`_
        for details).

    Returns
    -------
    pandas.DataFrame
        A dataframe where each item from the table matching the requested keys is
        represented by a row and its attributes by columns.

    Examples
    --------

    >>> df = get_df(
    ...     table="players",
    ...     keys=[{"player_id": "player_three"}, {"player_id": "player_one"}]
    ... )
    >>> print(df)
       bonus_points     player_id            last_play  rating        play_time
    0             4  player_three  2021-01-21 10:22:43     2.5  1 days 14:01:19
    1             3    player_one  2021-01-18 22:47:23     4.3  2 days 17:41:55

    By default, the data types of the returned dataframe are basic pandas/numpy types:

    >>> df.info()
    <class 'pandas.core.frame.DataFrame'>
    RangeIndex: 2 entries, 0 to 1
    Data columns (total 5 columns):
        #   Column        Non-Null Count  Dtype
       ---  ------        --------------  -----
        0   bonus_points  1 non-null      float64
        1   player_id     2 non-null      object
        2   last_play     2 non-null      object
        3   rating        2 non-null      float64
        4   play_time     2 non-null      object
    dtypes: float64(2), object(3)
    memory usage: 208.0+ bytes

    The ``dtype`` parameter can be used to specify the data types of the different
    columns:

    >>> df = get_df(
    ...     table="players",
    ...     keys=keys(player_id=["player_two", "player_four"]),
    ...     dtype={
    ...         "bonus_points": "Int8",
    ...         "last_play": "datetime64[ns, UTC]",
    ...         # "play_time": "timedelta64[ns]"  # See note below.
    ...     }
    ... )
    >>> df.info()
    <class 'pandas.core.frame.DataFrame'>
    RangeIndex: 2 entries, 0 to 1
    Data columns (total 5 columns):
        #   Column        Non-Null Count  Dtype
       ---  ------        --------------  -----
        0   bonus_points  1 non-null      Int8
        1   player_id     2 non-null      object
        2   last_play     2 non-null      datetime64[ns, UTC]
        3   rating        2 non-null      float64
        4   play_time     2 non-null      object
    dtypes: Int8(1), datetime64[ns, UTC](1), float64(1), object(2)
    memory usage: 196.0+ bytes

    .. note:: Due to a
        `known bug in pandas <https://github.com/pandas-dev/pandas/issues/38509>`_,
        timedelta strings cannot currently be converted back to timedelta64 type via the
        ``dtype`` parameter. Use the ``pandas.to_timedelta`` function instead:

        >>> df.play_time = pd.to_timedelta(df.play_time)
        >>> df.info()
        <class 'pandas.core.frame.DataFrame'>
        RangeIndex: 2 entries, 0 to 1
        Data columns (total 5 columns):
            #   Column        Non-Null Count  Dtype
           ---  ------        --------------  -----
            0   bonus_points  1 non-null      Int8
            1   player_id     2 non-null      object
            2   last_play     2 non-null      datetime64[ns, UTC]
            3   rating        2 non-null      float64
            4   play_time     2 non-null      timedelta64[ns]
        dtypes: Int8(1), datetime64[ns, UTC](1), float64(1), object(1), timedelta64[ns](1)
        memory usage: 196.0+ bytes

    Omitting the ``keys`` parameter performs a scan of the table and returns all the
    items.

    >>> df = get_df(table="players")
    >>> print(df)
        bonus_points     player_id            last_play  rating        play_time
    0           4.0  player_three  2021-01-21 10:22:43     2.5  1 days 14:01:19
    1           NaN   player_four  2021-01-22 13:51:12     4.8  0 days 03:45:49
    2           3.0    player_one  2021-01-18 22:47:23     4.3  2 days 17:41:55
    3           1.0    player_two  2021-01-19 19:07:54     3.8  0 days 22:07:34

    Specifying item attributes via the ``attributes`` parameter returns only the
    columns corresponding to the specified attributes:

    >>> df = get_df(table="players", attributes=["player_id", "rating"])
    >>> print(df)
          player_id  rating
    0  player_three     2.5
    1   player_four     4.8
    2    player_one     4.3
    3    player_two     3.8
    """  # noqa: E501
    if keys is not None:
        items = get_items(
            keys=keys, table=table, attributes=attributes, boto3_kwargs=boto3_kwargs
        )
    else:
        items = get_all_items(
            table=table, attributes=attributes, boto3_kwargs=boto3_kwargs
        )

    return _to_df(items=items, dtype=dtype)


def keys(**kwargs):
    """Generate a list of key dictionaries from the partition key attribute name and a
    list of values. This can simplify the generation of keys to use with the ``get_df``
    function when only a partition key is used.

    Parameters
    ----------
    **kwargs
        A single keyword argument corresponding to the partition key name with a value
        corresponding to the list of key values to return.

    Returns
    -------
    list[dict]
        A list of key dictionaries.

    Examples
    --------

    Assuming we have a table with ``player_id`` as the partition key, we can generate
    the list of keys from the list of players:

    >>> key_list = keys(player_id=["player_two", "player_three", "player_four"])
    >>> print(key_list)
    [{'player_id': 'player_one'}, {'player_id': 'player_three'}, {'player_id': 'player_four'}]
    """  # noqa: E501
    if len(kwargs.keys()) > 1:
        raise ValueError("Only one key attribute (partition key) is supported.")

    k = list(kwargs.keys())[0]
    return [{k: v} for v in kwargs[k]]


def put_df(df, *, table, boto3_kwargs={}):
    """Put rows of a dataframe as items into a table. If the item(s) do not exist in the
    table they are created, otherwise the existing items are replaced with the new ones.

    Parameters
    ----------
    df : pandas.DataFrame
        Dataframe of items to add/update in the table. The dataframe must, at a minimum,
        contain columns that correspond to the table's primary key attribute(s).

    table : str
        Name of the DynamoDB table.

    boto3_kwargs : dict
        Keyword arguments to pass to the underlying ``boto3.client('dynamodb')``
        function call (see `boto3 docs <https://boto3.amazonaws.com/v1/documentation/api/latest/reference/core/session.html#boto3.session.Session.client>`_
        for details).

    Examples
    --------
    Assume with have the following dataframe:

    >>> print(players_df)
          player_id           last_play       play_time  rating  bonus_points
    0    player_one 2021-01-18 22:47:23 2 days 17:41:55     4.3             3
    1    player_two 2021-01-19 19:07:54 0 days 22:07:34     3.8             1
    2  player_three 2021-01-21 10:22:43 1 days 14:01:19     2.5             4
    3   player_four 2021-01-22 13:51:12 0 days 03:45:49     4.8          <NA>

    The following will add or update the corresponding items in the table named
    ``players``:

    >>> put_df(players_df, table="players")
    """  # noqa: E501
    put_items(items=_to_items(df), table=table, boto3_kwargs=boto3_kwargs)


def _to_df(items, *, dtype=None):
    """Convert an item dictionary or list of item dictionaries into a pandas
    DataFrame."""
    if isinstance(items, dict):
        items = [items]

    df = pd.DataFrame(items)

    if dtype is not None:
        df = df.astype(dtype)

    return df


def _to_items(df):
    """Convert a pandas dataframe to a dictionary of items."""
    if not isinstance(df, pd.DataFrame):
        raise TypeError("df must be a pandas DataFrame")

    return df.to_dict("records")
